File: src/Uva/Grafos/11733-airports.cpp
#include <stdio.h>
#include <stdlib.h>
#include <vector>
#include <algorithm>

using namespace std;

#define MAX 100002
#define pii pair<int, int>
#define vdii vector<pair<int, pair<int, int> > >

int parent[MAX];

void reset_parents(int total_nodes) {
    for (int i = 0; i <= total_nodes; i++) {
        parent[i] = i;
    }
}

int find_parent(int a) {
    if (parent[a] == a)
        return a;
    return parent[a] = find_parent(parent[a]);
}

void union_set(int a, int b) {
    parent[find_parent(a)] = find_parent(b);
}

int count_parents(int total_nodes) {
    int count = 0;

    for (int i = 1; i <= total_nodes; i++) {
        if (parent[i] == i) {
            count ++;
        }
    }

    return count;
}

pii minimun_spanning_tree(int total_nodes, vdii nodes_distances, int max_cost) {
    reset_parents(total_nodes);
    int cost = 0;
    int extra_airports = 0;
    for (vdii:: iterator it = nodes_distances.begin(); it != nodes_distances.end(); it++) {
        int a = it->second.first;
        int b = it->second.second;
        if (find_parent(a) != find_parent(b)) {
            union_set(a, b);
            if (it->first < max_cost) {
                cost += it->first;
            } else {
                cost += max_cost;
                extra_airports ++;
            }
        }
    }
    return make_pair(cost, extra_airports);
}


int main(int argc, char const *argv[]) {
    int ncases, nroads, nlocations, airport_cost;

    scanf("%d", &ncases);

    for (int icase = 1; icase <= ncases; icase++) {
        scanf("%d %d %d", &nlocations, &nroads, &airport_cost);

        vdii roads;
        for (int i = 0; i < nroads; ++i) {
            int location_i, location_j, road_cost;
            scanf("%d %d %d", &location_i, &location_j, &road_cost);
            roads.push_back(make_pair(road_cost, make_pair(location_i, location_j)));
        }

        sort(roads.begin(), roads.end());

        pii result = minimun_spanning_tree(nlocations, roads, airport_cost);
        int components = count_parents(nlocations);
        int cost = result.first + components * airport_cost;
        printf("Case #%d: %d %d\n", icase, cost , result.second + components);
    }

    return 0;
}

File: src/Uva/Grafos/11733-airports.py
import sys

parents = []


def read_line(split_func=None):
    line =  sys.stdin.readline().strip()
    if split_func:
        return [split_func(x) for x in line.split(' ')]
    return line


def reset_parents(size):
    global parents
    parents = [i for i in range(size + 1)]


def find_parent(a):
    if parents[a] == a:
        return a
    parents[a] = find_parent(parents[a])
    return parents[a]


def union_set(a, b):
    parents[find_parent(a)] = find_parent(b)


def count_componentes(size):
    count = 0
    for i in range(1, size + 1):
        if parents[i] == i:
            count += 1
    return count


def minimum_spanning_tree(roads, airport_cost):
    minimun_cost = 0
    extra_airports = 0
    for road_cost, node_i, node_j in roads:
        if find_parent(node_i) != find_parent(node_j):
            union_set(node_i, node_j)
            if road_cost < airport_cost:
                minimun_cost += road_cost
            else:
                minimun_cost += airport_cost
                extra_airports += 1

    return minimun_cost, extra_airports


def main():
    ncases = int(read_line())

    for case in range(ncases):
        nlocations, nroads, airport_cost = read_line(split_func=int)

        roads = []
        for i in range(nroads):
            location_i, location_j, road_cost = read_line(split_func=int)
            roads.append((road_cost, location_i, location_j))

        roads.sort()

        reset_parents(nlocations)
        cost, extra_airports = minimum_spanning_tree(roads, airport_cost)
        components = count_componentes(nlocations)

        cost += components * airport_cost
        nairports = components + extra_airports

        print('Case %d: %d %d' % (case + 1, cost, nairports))


if __name__ == '__main__':
    main()
